File: packages/validator-ts/src/validate/parameters.ts
/**
 * Rule parameter checks (validation-rules.md, "Parameter errors").
 *
 * A parameter outside the definitions, and a rule name that is not registered, is a
 * load failure. Declared rule names and parameters are
 * checked after composition and the forbidden-key scan, fields in declaration
 * order and each field's rules in declaration order; a parameter selected by a
 * condition is checked when the validator selects it.
 */

import { ComposeLoadError } from '../compose/errors';
import { isFiniteNumber, isLengthLimit, isLengthRange, isNumberRange, isStep, readMembers } from '../values/index';
import { PatternSyntaxError, compilePattern } from '../pattern/index';
import { getRule } from '../rules/index';

/** The failure a rule parameter causes. */
export interface ParameterFailure {
  code: 'INVALID_RULE_PARAMETER' | 'INVALID_RULE_PATTERN';
  message: string;
}

/** Rule names whose parameters this module checks. */
const CHECKED_RULES = new Set([
  'minlength',
  'maxlength',
  'rangelength',
  'number',
  'digits',
  'min',
  'max',
  'range',
  'step',
  'mincount',
  'maxcount',
  'in',
  'match',
  'pattern',
]);

/**
 * The failure an effective (resolved) parameter of a rule causes, or `null`.
 * `false` and `null` disable a rule and are never failures.
 */
export function ruleParameterFailure(ruleName: string, param: unknown): ParameterFailure | null {
  if (param === false || param === null || !CHECKED_RULES.has(ruleName)) {
    return null;
  }
  const invalid = (message: string): ParameterFailure => ({ code: 'INVALID_RULE_PARAMETER', message });
  switch (ruleName) {
    case 'minlength':
    case 'maxlength':
    case 'mincount':
    case 'maxcount':
      return isLengthLimit(param)
        ? null
        : invalid(`Invalid ${ruleName} parameter: expected an integer from 0 to 9007199254740991`);
    case 'rangelength':
      return isLengthRange(param)
        ? null
        : invalid('Invalid rangelength parameter: expected [minimum, maximum] integers with minimum not above maximum');
    case 'number':
    case 'digits':
      return param === true ? null : invalid(`Invalid ${ruleName} parameter: expected true or false`);
    case 'min':
    case 'max':
      return isFiniteNumber(param) ? null : invalid(`Invalid ${ruleName} parameter: expected a finite number`);
    case 'range':
      return isNumberRange(param)
        ? null
        : invalid('Invalid range parameter: expected [minimum, maximum] finite numbers with minimum not above maximum');
    case 'step':
      return isStep(param) ? null : invalid('Invalid step parameter: expected a finite number above 0');
    case 'in': {
      const members = readMembers(param);
      return 'error' in members ? invalid(members.error) : null;
    }
    default: {
      if (typeof param !== 'string') return invalid(`Invalid ${ruleName} parameter: expected a pattern string`);
      try {
        // Compiled once here, when the parameter is checked.
        compilePattern(param);
        return null;
      } catch (error) {
        if (!(error instanceof PatternSyntaxError)) throw error;
        return {
          code: 'INVALID_RULE_PATTERN',
          message: `Invalid ${ruleName} pattern: ${error.reason} at ${error.offset}`,
        };
      }
    }
  }
}

/**
 * Throw the load failure of an effective parameter.
 *
 * @param declarationPath the field's property names from the root, without row keys.
 * @throws {ComposeLoadError} when the parameter is outside the definitions.
 */
export function assertRuleParameter(ruleName: string, param: unknown, declarationPath: string[]): void {
  const failure = ruleParameterFailure(ruleName, param);
  if (failure) throw new ComposeLoadError(failure.code, failure.message, [...declarationPath]);
}

/**
 * Throw the load failure of a `validate` or `messages` key that is not a registered rule.
 *
 * @param declarationPath the field's property names from the root, without row keys.
 * @throws {ComposeLoadError} `UNKNOWN_RULE` when the name is not registered.
 */
export function assertRuleName(ruleName: string, declarationPath: string[]): void {
  if (getRule(ruleName) === undefined) {
    throw new ComposeLoadError('UNKNOWN_RULE', `Unknown rule: ${ruleName}`, [...declarationPath]);
  }
}

File: packages/validator-rust/src/validate/parameters.rs
//! Rule parameter checks (validation rules, "Parameter errors").
//!
//! A parameter outside the definitions is a load failure located at the field's
//! declaration path. Parameters are checked when the specification loads, fields
//! and rules in declaration order, including every literal a condition map or a
//! ternary can select; a value a ternary takes from the data is checked when it is
//! selected. Checking yields the typed parameter the rule uses.

use std::collections::HashMap;
use std::sync::Arc;

use serde_json::{Map, Value};

use crate::compose::{ComposeErrorCode, ComposeLoadError};

use super::errors::ValidateError;
use super::length::{limit, range_limits, MAX_LIMIT};
use super::membership::{members, Comparable};
use super::pattern::{compile, Program};
use super::validator::{
    is_conditional, is_disabled, is_group_with_properties, selectable_literals,
};

/// A checked rule parameter.
pub(crate) enum Parameter {
    /// A rule whose parameter needs no check here.
    Unchecked,
    /// A `minlength` or `maxlength` limit.
    Limit(u64),
    /// `rangelength` limits.
    Range(u64, u64),
    /// A `min` or `max` bound.
    Bound(f64),
    /// `range` bounds.
    Bounds(f64, f64),
    /// A `step` above 0.
    Step(f64),
    /// A `mincount` or `maxcount` limit.
    Count(u64),
    /// `in` members.
    Members(Vec<Comparable>),
    /// A compiled `pattern` or `match`.
    Pattern(Arc<Program>),
}

/// Patterns compiled while the specification loaded, by source text.
#[derive(Default)]
pub(crate) struct Patterns(HashMap<String, Arc<Program>>);

/// Check every declared parameter of `properties`; returns the compiled patterns.
pub(crate) fn check_declared(properties: &Map<String, Value>) -> Result<Patterns, ValidateError> {
    let mut patterns = Patterns::default();
    check_fields(properties, &mut Vec::new(), &mut patterns)?;
    Ok(patterns)
}

fn check_fields(
    properties: &Map<String, Value>,
    path: &mut Vec<String>,
    patterns: &mut Patterns,
) -> Result<(), ValidateError> {
    for (key, field) in properties {
        if !field.is_object() {
            continue;
        }
        path.push(key.clone());
        if let Some(Value::Object(rules)) = field.get("validate") {
            for (rule, value) in rules {
                if is_disabled(value) {
                    continue;
                }
                if is_conditional(rule, value) {
                    for literal in selectable_literals(rule, value) {
                        if !is_disabled(&literal) {
                            check(rule, &literal, path, patterns)?;
                        }
                    }
                    continue;
                }
                if let Parameter::Pattern(program) = check(rule, value, path, patterns)? {
                    if let Value::String(source) = value {
                        patterns.0.insert(source.clone(), program);
                    }
                }
            }
        }
        if let Some(children) = is_group_with_properties(field) {
            check_fields(children, path, patterns)?;
        }
        path.pop();
    }
    Ok(())
}

/// Check the parameter `rule` uses; `path` is the field's declaration path.
/// A `false` or `null` parameter disables the rule and is never passed here.
pub(crate) fn check(
    rule: &str,
    parameter: &Value,
    path: &[String],
    patterns: &Patterns,
) -> Result<Parameter, ValidateError> {
    let invalid = |message: String| {
        ValidateError::Load(ComposeLoadError::with_trace(
            ComposeErrorCode::InvalidRuleParameter,
            message,
            path.to_vec(),
        ))
    };
    match rule {
        "minlength" | "maxlength" => limit(parameter).map(Parameter::Limit).ok_or_else(|| {
            invalid(format!(
                "Invalid {rule} parameter: expected an integer from 0 to {MAX_LIMIT}"
            ))
        }),
        "rangelength" => range_limits(parameter)
            .map(|(minimum, maximum)| Parameter::Range(minimum, maximum))
            .ok_or_else(|| {
                invalid(
                    "Invalid rangelength parameter: expected [minimum, maximum] integers \
                     with minimum not above maximum"
                        .to_string(),
                )
            }),
        "number" | "digits" => match parameter {
            Value::Bool(_) => Ok(Parameter::Unchecked),
            _ => Err(invalid(format!(
                "Invalid {rule} parameter: expected true or false"
            ))),
        },
        "min" | "max" => finite(parameter).map(Parameter::Bound).ok_or_else(|| {
            invalid(format!(
                "Invalid {rule} parameter: expected a finite number"
            ))
        }),
        "range" => match parameter.as_array().map(Vec::as_slice) {
            Some([minimum, maximum]) => match (finite(minimum), finite(maximum)) {
                (Some(minimum), Some(maximum)) if minimum <= maximum => {
                    Some(Parameter::Bounds(minimum, maximum))
                }
                _ => None,
            },
            _ => None,
        }
        .ok_or_else(|| {
            invalid(
                "Invalid range parameter: expected [minimum, maximum] finite numbers \
                 with minimum not above maximum"
                    .to_string(),
            )
        }),
        "step" => finite(parameter)
            .filter(|step| *step > 0.0)
            .map(Parameter::Step)
            .ok_or_else(|| {
                invalid("Invalid step parameter: expected a finite number above 0".to_string())
            }),
        "mincount" | "maxcount" => limit(parameter).map(Parameter::Count).ok_or_else(|| {
            invalid(format!(
                "Invalid {rule} parameter: expected an integer from 0 to {MAX_LIMIT}"
            ))
        }),
        "in" => members(parameter)
            .map(Parameter::Members)
            .map_err(|error| invalid(error.message().to_string())),
        "pattern" | "match" => {
            let Value::String(source) = parameter else {
                return Err(invalid(format!(
                    "Invalid {rule} parameter: expected a pattern string"
                )));
            };
            if let Some(program) = patterns.0.get(source) {
                return Ok(Parameter::Pattern(Arc::clone(program)));
            }
            compile(source)
                .map(|program| Parameter::Pattern(Arc::new(program)))
                .map_err(|error| {
                    ValidateError::Load(ComposeLoadError::with_trace(
                        ComposeErrorCode::InvalidRulePattern,
                        format!("Invalid {rule} pattern: {error}"),
                        path.to_vec(),
                    ))
                })
        }
        _ => Ok(Parameter::Unchecked),
    }
}

/// A JSON number as a finite double.
fn finite(value: &Value) -> Option<f64> {
    value.as_f64().filter(|number| number.is_finite())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn failure(properties: Value) -> (String, String, String) {
        let properties = properties.as_object().cloned().expect("a properties map");
        let Err(error) = check_declared(&properties) else {
            panic!("the declaration loaded");
        };
        (
            error.code().to_string(),
            error.message().to_string(),
            error.at(),
        )
    }

    #[test]
    fn fields_and_rules_are_checked_in_declaration_order() {
        let (code, message, at) = failure(json!({
            "a": { "type": "text", "validate": { "in": [], "pattern": "(" } },
            "b": { "type": "text", "validate": { "maxlength": -1 } },
        }));
        assert_eq!(code, "INVALID_RULE_PARAMETER");
        assert_eq!(message, "Invalid in parameter: members must not be empty");
        assert_eq!(at, "a");

        let (code, message, at) = failure(json!({
            "g": {
                "type": "group",
                "multiple": true,
                "validate": { "required": true },
                "properties": { "n": { "type": "text", "validate": { "match": "a{2}{3}" } } },
            },
        }));
        assert_eq!(code, "INVALID_RULE_PATTERN");
        assert_eq!(message, "Invalid match pattern: invalid quantifier at 4");
        assert_eq!(at, "g.n");
    }

    #[test]
    fn selectable_literals_are_checked_at_load() {
        for (rule, value, message) in [
            (
                "minlength",
                json!({ ".big": 1.5, "true": 2 }),
                "Invalid minlength parameter: expected an integer from 0 to 9007199254740991",
            ),
            (
                "maxlength",
                json!(".big ? 3 : -1"),
                "Invalid maxlength parameter: expected an integer from 0 to 9007199254740991",
            ),
            (
                "maxlength",
                json!(".a ? .limit : .b ? 2 : \"x\""),
                "Invalid maxlength parameter: expected an integer from 0 to 9007199254740991",
            ),
            (
                "rangelength",
                json!({ ".a": [3, 2] }),
                "Invalid rangelength parameter: expected [minimum, maximum] integers \
                 with minimum not above maximum",
            ),
        ] {
            let (code, actual, at) = failure(json!({
                "g": { "type": "group", "properties": {
                    "v": { "type": "text", "validate": { rule: value } },
                } },
            }));
            assert_eq!(code, "INVALID_RULE_PARAMETER");
            assert_eq!(actual, message, "{value}");
            assert_eq!(at, "g.v");
        }
    }

    #[test]
    fn disabled_conditional_and_unchecked_parameters_load() {
        let properties = json!({
            "a": { "type": "text", "validate": {
                "minlength": false, "maxlength": null, "in": false, "pattern": null,
                "rangelength": ".x ? .range : false", "match": false,
                "min": ".x", "unknown": [],
            } },
            "b": { "type": "text", "validate": {
                "maxlength": { ".x": 3, "true": null },
                "minlength": ".x ? .limit : .y ? 2 : false",
            } },
            // Only groups are descended into, as the validator does.
            "c": { "type": "text", "properties": { "d": { "validate": { "in": [] } } } },
        });
        if let Err(error) = check_declared(properties.as_object().unwrap()) {
            panic!("{error:?}");
        }
    }
}
